airbyte-integrations/bases/debezium/src/main/java/io/airbyte/integrations/debezium/internals/DebeziumShutdownProcedure.java

```diff
@@ -0,0 +1,126 @@
+/*
+ * Copyright (c) 2023 Airbyte, Inc., all rights reserved.
+ */
+
+package io.airbyte.integrations.debezium.internals;
+
+import io.airbyte.commons.concurrency.VoidCallable;
+import io.airbyte.commons.lang.MoreBooleans;
+import java.util.concurrent.ExecutorService;
+import java.util.concurrent.Executors;
+import java.util.concurrent.LinkedBlockingQueue;
+import java.util.concurrent.TimeUnit;
+import java.util.function.Supplier;
+import org.slf4j.Logger;
+import org.slf4j.LoggerFactory;
+
+/**
+ * This class has the logic for shutting down Debezium Engine in graceful manner. We made it Generic
+ * to allow us to write tests easily.
+ */
+public class DebeziumShutdownProcedure<T> {
+
+  private static final Logger LOGGER = LoggerFactory.getLogger(DebeziumShutdownProcedure.class);
+  private final LinkedBlockingQueue<T> sourceQueue;
+  private final LinkedBlockingQueue<T> targetQueue;
+  private final ExecutorService executorService;
+  private final Supplier<Boolean> publisherStatusSupplier;
+  private final VoidCallable debeziumThreadRequestClose;
+  private Throwable exception;
+  private boolean hasTransferThreadShutdown;
+
+  public DebeziumShutdownProcedure(final LinkedBlockingQueue<T> sourceQueue,
+                                   final VoidCallable debeziumThreadRequestClose,
+                                   final Supplier<Boolean> publisherStatusSupplier) {
+    this.sourceQueue = sourceQueue;
+    this.targetQueue = new LinkedBlockingQueue<>();
+    this.debeziumThreadRequestClose = debeziumThreadRequestClose;
+    this.publisherStatusSupplier = publisherStatusSupplier;
+    this.hasTransferThreadShutdown = false;
+    this.executorService = Executors.newSingleThreadExecutor(r -> {
+      final Thread thread = new Thread(r, "queue-data-transfer-thread");
+      thread.setUncaughtExceptionHandler((t, e) -> {
+        exception = e;
+      });
+      return thread;
+    });
+  }
+
+  private Runnable transfer() {
+    return () -> {
+      while (!sourceQueue.isEmpty() || !MoreBooleans.isTruthy(publisherStatusSupplier.get())) {
+        try {
+          T event = sourceQueue.poll(10, TimeUnit.SECONDS);
+          if (event != null) {
+            targetQueue.put(event);
```

Comment:
Suggestion: use offer to avoid blocking / error in case target queue is at capacity

Reply:
The targetQueue is not bound by size so it should not be at capacity.